File: src/commands/wr.ts
import { SlashCommandBuilder, ChatInputCommandInteraction, EmbedBuilder } from "discord.js";
import { reply } from "../lib/functions/discord";
import { parseTime } from "../lib/functions/util";
import { getMaps, getWR, validateMap } from "gokz.js";
import userSchema from "../lib/schemas/user";
import modeMap from "gokz.js/lib/api";
import SchnoseBot from "src/classes/Schnose";

export default {
	data: new SlashCommandBuilder()
		.setName("wr")
		.setDescription("Check the World Record on a map.")
		.addStringOption((o) => o.setName("map").setDescription("Specify a map.").setRequired(true))
		.addStringOption((o) =>
			o.setName("mode").setDescription("Specify a mode.").setChoices(
				{
					name: "KZT",
					value: "kz_timer"
				},
				{
					name: "SKZ",
					value: "kz_simple"
				},
				{
					name: "VNL",
					value: "kz_vanilla"
				}
			)
		),

	async execute(interaction: ChatInputCommandInteraction, client: SchnoseBot) {
		await interaction.deferReply();

		const inputMap = interaction.options.getString("map")!;
		const inputMode = interaction.options.getString("mode") || null;

		const globalMaps = await getMaps();
		if (!globalMaps.success) return reply(interaction, { content: globalMaps.error });

		const mapValidation = await validateMap(inputMap, globalMaps.data!);
		if (!mapValidation.success) return reply(interaction, { content: mapValidation.error });

		let mode: string;
		if (inputMode) mode = inputMode;
		else {
			const userDB = await userSchema.find({ discordID: interaction.user.id });
			if (!userDB[0]?.mode)
				return reply(interaction, {
					content: "You must either specify a mode or set a default value using `/mode`."
				});
			else mode = userDB[0].mode;
		}

		const req = await Promise.all([
			await getWR(mapValidation.data!.name, 0, mode, true),
			await getWR(mapValidation.data!.name, 0, mode, false)
		]);

		const embed = new EmbedBuilder()
			.setColor([116, 128, 194])
			.setTitle(
				`[WR] - ${mapValidation.data!.name} (${modeMap.get(
					req[0].data?.mode || req[1].data?.mode
				)})`
			)
			.setURL(`https://kzgo.eu/maps/${mapValidation.data!.name}`)
			.setThumbnail(
				`https://raw.githubusercontent.com/KZGlobalTeam/map-images/master/images/${
					mapValidation.data!.name
				}.jpg`
			)
			.addFields([
				{
					name: "TP",
					value: `${parseTime(req[0].data?.time || 0)} (${req[0].data?.player_name || "-"})`,
					inline: true
				},
				{
					name: "PRO",
					value: `${parseTime(req[1].data?.time || 0)} (${req[1].data?.player_name || "-"})`,
					inline: true
				}
			])
			.setFooter({
				text: "(͡ ͡° ͜ つ ͡͡°)7",
				iconURL: client.icon
			});

		return reply(interaction, {
			embeds: [embed]
		});
	}
};

File: src/commands/wr.rs
use {
	crate::{
		events::slash_commands::{InteractionState, InteractionResponseData::*},
		schnose::InteractionResult,
		db::retrieve_mode,
		util::*,
		gokz,
	},
	futures::future::join_all,
	gokz_rs::{prelude::*, global_api::*},
	itertools::Itertools,
	serenity::{
		builder::{CreateApplicationCommand, CreateEmbed},
		model::prelude::command::CommandOptionType,
	},
};

pub(crate) fn register(cmd: &mut CreateApplicationCommand) -> &mut CreateApplicationCommand {
	return cmd
		.name("wr")
		.description("Check the World Record on a map.")
		.create_option(|opt| {
			opt.kind(CommandOptionType::String)
				.name("map_name")
				.description("Specify a map.")
				.required(true)
		})
		.create_option(|opt| {
			opt.kind(CommandOptionType::String)
				.name("mode")
				.description("Choose a mode.")
				.add_string_choice("KZT", "kz_timer")
				.add_string_choice("SKZ", "kz_simple")
				.add_string_choice("VNL", "kz_vanilla")
				.required(false)
		});
}

pub(crate) async fn execute(state: &mut InteractionState<'_>) -> InteractionResult {
	// Defer current interaction since this could take a while
	state.defer().await?;

	let map_name = state.get::<String>("map_name").expect("This option is marked as `required`.");

	let mode = match state.get::<String>("mode") {
		Some(mode_name) => Mode::from_str(&mode_name)
			.expect("The possible values for this are hard-coded and should never be invalid."),
		None => retrieve_mode(state.user, state.db).await?,
	};

	let global_maps = get_maps(&state.req_client).await?;

	let map = is_global(&MapIdentifier::Name(map_name), &global_maps).await?;

	let map_identifier = MapIdentifier::Name(map.name.clone());

	let (tp, pro) = join_all([
		get_wr(&map_identifier, &mode, true, 0, &state.req_client),
		get_wr(&map_identifier, &mode, false, 0, &state.req_client),
	])
	.await
	.into_iter()
	.collect_tuple()
	.expect("This cannot fail, look 6 lines up.");

	if let (&Err(_), &Err(_)) = (&tp, &pro) {
		return Ok(Message("No WRs found 😔.".into()));
	}

	let links = (gokz::get_replay_link(&tp).await, gokz::get_replay_link(&pro).await);

	let mut embed = CreateEmbed::default()
		.colour(state.colour)
		.title(format!("[WR] {} (T{})", &map.name, &map.difficulty))
		.url(format!(
			"https://kzgo.eu/maps/{}?{}=",
			&map.name,
			&mode.to_fancy().to_lowercase()
		))
		.thumbnail(state.thumbnail(&map.name))
		.field(
			"TP",
			format!(
				"{} ({})",
				match &tp {
					Ok(rec) => format_time(rec.time),
					_ => String::from("😔"),
				},
				{
					let mut name = "unknown";
					if let Ok(rec) = &tp {
						if let Some(player_name) = &rec.player_name {
							name = &player_name;
						}
					}
					name
				}
			),
			true,
		)
		.field(
			"PRO",
			format!(
				"{} ({})",
				match &pro {
					Ok(rec) => format_time(rec.time),
					_ => String::from("😔"),
				},
				{
					let mut name = "unknown";
					if let Ok(rec) = &pro {
						if let Some(player_name) = &rec.player_name {
							name = &player_name;
						}
					}
					name
				}
			),
			true,
		)
		.footer(|f| f.text(format!("Mode: {}", mode.to_fancy())).icon_url(&state.icon))
		.to_owned();

	attach_replay_links(&mut embed, links);

	return Ok(Embed(embed));
}
